File: ai-content/ollama_client.py
"""Reusable Ollama adapter for Captain's Log rewrite pipeline."""

from __future__ import annotations

import json
import time
from typing import Any

import requests

from ai_output_quality import validate_ai_output
from config import (
    OLLAMA_MAX_RETRIES,
    OLLAMA_MODEL,
    OLLAMA_NUM_PREDICT,
    OLLAMA_STREAM_TIMEOUT_SEC,
    OLLAMA_TEMPERATURE,
    OLLAMA_TOP_P,
    OLLAMA_URL,
)
from pipeline_metrics import record_if_requests_timeout


def _resolve_base_url() -> str:
    raw = (OLLAMA_URL or "http://127.0.0.1:11434/api/generate").strip()
    return raw.replace("/api/generate", "").replace("/api/chat", "").rstrip("/")


def _resolve_model(override: str | None = None) -> str:
    return (override or OLLAMA_MODEL or "llama3.1:8b").strip() or "llama3.1:8b"


def _resolve_num_predict(override: int | None = None) -> int:
    if override and override > 0:
        return int(override)
    if OLLAMA_NUM_PREDICT and OLLAMA_NUM_PREDICT > 0:
        return int(OLLAMA_NUM_PREDICT)
    return 2400


def _resolve_temperature(override: float | None = None) -> float:
    if override is not None:
        return max(0.05, min(0.95, float(override)))
    return max(0.05, min(0.95, float(OLLAMA_TEMPERATURE)))


def _stream_generate(payload: dict[str, Any], timeout_sec: int) -> str:
    url = f"{_resolve_base_url()}/api/generate"
    parts: list[str] = []
    with requests.post(
        url,
        json=payload,
        stream=True,
        timeout=timeout_sec,
        headers={"Content-Type": "application/json"},
    ) as response:
        response.raise_for_status()
        for line in response.iter_lines():
            if not line:
                continue
            try:
                chunk = json.loads(line.decode("utf-8"))
            except json.JSONDecodeError:
                continue
            parts.append(chunk.get("response", "") or "")
    return "".join(parts).strip()


def generate_text(
    prompt: str,
    *,
    system: str | None = None,
    model: str | None = None,
    temperature: float | None = None,
    num_predict: int | None = None,
    timeout_sec: int | None = None,
    stream: bool = True,
    format_json: bool = False,
) -> tuple[bool, str, str | None]:
    """Return (ok, text, error)."""
    payload: dict[str, Any] = {
        "model": _resolve_model(model),
        "prompt": str(prompt or ""),
        "stream": stream,
        "options": {
            "num_predict": _resolve_num_predict(num_predict),
            "temperature": _resolve_temperature(temperature),
            "top_p": float(OLLAMA_TOP_P),
        },
    }
    if system:
        payload["system"] = system
    if format_json:
        payload["format"] = "json"

    timeout = timeout_sec if timeout_sec else OLLAMA_STREAM_TIMEOUT_SEC
    attempts = max(1, int(OLLAMA_MAX_RETRIES))
    last_err: str | None = None

    for attempt in range(1, attempts + 1):
        try:
            if stream:
                text = _stream_generate(payload, timeout)
            else:
                url = f"{_resolve_base_url()}/api/generate"
                r = requests.post(url, json={**payload, "stream": False}, timeout=timeout)
                r.raise_for_status()
                data = r.json()
                text = str(data.get("response") or "").strip()
            if text:
                return True, text, None
            last_err = "empty_response"
        except Exception as e:
            record_if_requests_timeout(e)
            last_err = str(e)
            print(f"[ollama_client] attempt {attempt}/{attempts} failed: {e}")
        if attempt < attempts:
            time.sleep(1.5 * attempt)
    return False, "", last_err or "generation_failed"


def generate_validated(
    prompt: str,
    *,
    system: str | None = None,
    model: str | None = None,
    temperature: float | None = None,
    num_predict: int | None = None,
    timeout_sec: int | None = None,
    min_words: int = 80,
    title: str = "",
    source_text: str = "",
    required_sections: list[str] | None = None,
) -> tuple[bool, str, str | None, dict[str, Any]]:
    """Generate with one stricter retry; returns (ok, text, error, meta)."""
    meta: dict[str, Any] = {}
    ok, text, err = generate_text(
        prompt,
        system=system,
        model=model,
        temperature=temperature,
        num_predict=num_predict,
        timeout_sec=timeout_sec,
    )
    if not ok or not text:
        return False, "", err or "generation_failed", meta

    qc_ok, reason, qc_meta = validate_ai_output(
        text,
        min_words=min_words,
        title=title,
        source_text=source_text,
        prompt=prompt,
        required_sections=required_sections,
    )
    meta.update(qc_meta or {})
    if qc_ok:
        return True, text, None, meta

    strict_prompt = (
        f"{prompt}\n\nSTRICT RETRY: Paraphrase SOURCE FACTS only. "
        "Do not repeat the headline. No placeholder phrases. Follow required headings exactly."
    )
    ok2, text2, err2 = generate_text(
        strict_prompt,
        system=system,
        model=model,
        temperature=max(0.05, _resolve_temperature(temperature) - 0.08),
        num_predict=num_predict,
        timeout_sec=timeout_sec,
    )
    if not ok2 or not text2:
        return False, "", err2 or reason, meta

    qc_ok2, reason2, qc_meta2 = validate_ai_output(
        text2,
        min_words=min_words,
        title=title,
        source_text=source_text,
        prompt=prompt,
        required_sections=required_sections,
    )
    meta.update(qc_meta2 or {})
    if qc_ok2:
        meta["retried"] = True
        return True, text2, None, meta
    return False, "", reason2 or "quality_rejected", meta

File: server/services/ollamaClient.js
/**
 * Reusable Ollama adapter: timeouts, retries, generation options, optional quality validation.
 * Used by rocket advisories, Captain Bobby, and (via mirrored Python client) Captain's Log.
 */

const fetch = require('node-fetch');
const { validateAiOutput } = require('./aiOutputQuality');

function trimTrailingSlash(url) {
  return String(url || '').replace(/\/$/, '');
}

function resolveOllamaBaseUrl() {
  const raw = String(process.env.OLLAMA_URL || 'http://127.0.0.1:11434').trim();
  if (!raw) return 'http://127.0.0.1:11434';
  const cleaned = raw.replace(/\/api\/generate\/?$/i, '').replace(/\/api\/chat\/?$/i, '');
  return trimTrailingSlash(cleaned);
}

function resolveModel() {
  return (
    String(process.env.CAPTAIN_BOBBY_MODEL || process.env.OLLAMA_MODEL || 'llama3.1:8b').trim() ||
    'llama3.1:8b'
  );
}

function resolveTimeoutMs(override) {
  if (Number.isFinite(Number(override)) && Number(override) > 0) return Number(override);
  const fromEnv = Number(process.env.OLLAMA_TIMEOUT_MS || process.env.OLLAMA_TIMEOUT_SEC * 1000 || 90000);
  return Number.isFinite(fromEnv) && fromEnv > 0 ? fromEnv : 90000;
}

function resolveNumPredict(override) {
  if (Number.isFinite(Number(override)) && Number(override) > 0) return Math.floor(Number(override));
  const bobby = Number(process.env.CAPTAIN_BOBBY_MAX_TOKENS || 0);
  const shared = Number(process.env.OLLAMA_NUM_PREDICT || 0);
  const chosen = bobby > 0 ? bobby : shared;
  return chosen > 0 ? Math.floor(chosen) : 800;
}

function resolveTemperature(override) {
  if (Number.isFinite(Number(override))) return Math.max(0.05, Math.min(1, Number(override)));
  const raw = Number(process.env.OLLAMA_TEMPERATURE || 0.2);
  return Number.isFinite(raw) ? Math.max(0.05, Math.min(0.95, raw)) : 0.2;
}

function isEnabled() {
  return String(process.env.CAPTAIN_BOBBY_ENABLED || '1').trim() !== '0';
}

async function fetchWithTimeout(url, options, timeoutMs) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    return await fetch(url, { ...options, signal: controller.signal });
  } finally {
    clearTimeout(timer);
  }
}

/**
 * @param {object} params
 * @param {string} params.prompt
 * @param {string} [params.system]
 * @param {string} [params.model]
 * @param {number} [params.temperature]
 * @param {number} [params.numPredict]
 * @param {number} [params.timeoutMs]
 * @param {string} [params.format] json | undefined
 * @param {boolean} [params.stream=false]
 * @returns {Promise<{ ok: boolean, text: string, error?: string, raw?: unknown }>}
 */
async function generateText(params = {}) {
  if (!isEnabled()) {
    return { ok: false, text: '', error: 'ollama_disabled' };
  }

  const base = resolveOllamaBaseUrl();
  const url = `${base}/api/generate`;
  const model = params.model || resolveModel();
  const timeoutMs = resolveTimeoutMs(params.timeoutMs);
  const body = {
    model,
    prompt: String(params.prompt || ''),
    stream: Boolean(params.stream),
    options: {
      temperature: resolveTemperature(params.temperature),
      top_p: Number(process.env.OLLAMA_TOP_P || 0.85),
      num_predict: resolveNumPredict(params.numPredict),
    },
  };
  if (params.system) {
    body.system = String(params.system);
  }
  if (params.format === 'json') {
    body.format = 'json';
  }

  try {
    const response = await fetchWithTimeout(
      url,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
      },
      timeoutMs
    );
    const rawText = await response.text();
    let data;
    try {
      data = JSON.parse(rawText);
    } catch {
      return { ok: false, text: '', error: 'invalid_json', raw: rawText.slice(0, 240) };
    }
    if (!response.ok) {
      return { ok: false, text: '', error: `http_${response.status}`, raw: data };
    }
    const text = typeof data.response === 'string' ? data.response.trim() : '';
    if (!text) {
      return { ok: false, text: '', error: 'empty_response', raw: data };
    }
    return { ok: true, text, raw: data };
  } catch (err) {
    const message = err?.name === 'AbortError' ? 'timeout' : err?.message || 'request_failed';
    return { ok: false, text: '', error: message };
  }
}

/**
 * Chat-style generation for Captain Bobby.
 * @param {object} params
 * @param {Array<{role: string, content: string}>} params.messages
 * @param {string} [params.system]
 */
async function generateChat(params = {}) {
  if (!isEnabled()) {
    return { ok: false, text: '', error: 'ollama_disabled' };
  }

  const base = resolveOllamaBaseUrl();
  const url = `${base}/api/chat`;
  const model = params.model || resolveModel();
  const timeoutMs = resolveTimeoutMs(params.timeoutMs);
  const messages = Array.isArray(params.messages) ? params.messages : [];
  const chatMessages = params.system
    ? [{ role: 'system', content: String(params.system) }, ...messages]
    : messages;

  const body = {
    model,
    messages: chatMessages,
    stream: false,
    options: {
      temperature: resolveTemperature(params.temperature),
      top_p: Number(process.env.OLLAMA_TOP_P || 0.85),
      num_predict: resolveNumPredict(params.numPredict),
    },
  };
  if (params.format === 'json') {
    body.format = 'json';
  }

  try {
    const response = await fetchWithTimeout(
      url,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
      },
      timeoutMs
    );
    const rawText = await response.text();
    let data;
    try {
      data = JSON.parse(rawText);
    } catch {
      return { ok: false, text: '', error: 'invalid_json', raw: rawText.slice(0, 240) };
    }
    if (!response.ok) {
      return { ok: false, text: '', error: `http_${response.status}`, raw: data };
    }
    const text =
      typeof data.message?.content === 'string'
        ? data.message.content.trim()
        : typeof data.response === 'string'
          ? data.response.trim()
          : '';
    if (!text) {
      return { ok: false, text: '', error: 'empty_response', raw: data };
    }
    return { ok: true, text, raw: data };
  } catch (err) {
    const message = err?.name === 'AbortError' ? 'timeout' : err?.message || 'request_failed';
    return { ok: false, text: '', error: message };
  }
}

/**
 * Generate with one quality-validated retry using a stricter temperature.
 * @param {object} params
 * @param {object} [params.quality]
 */
async function generateValidated(params = {}) {
  const quality = params.quality || {};
  const first = await generateText(params);
  if (!first.ok) return first;

  let check = validateAiOutput(first.text, quality);
  if (check.ok) {
    return { ...first, quality: check.meta };
  }

  const retryPrompt = `${String(params.prompt || '')}\n\nSTRICT RETRY: Be concise, factual, and avoid repeating the headline. No placeholder phrases.`;
  const retry = await generateText({
    ...params,
    prompt: retryPrompt,
    temperature: Math.max(0.05, resolveTemperature(params.temperature) - 0.08),
  });
  if (!retry.ok) {
    return { ...first, qualityRejected: true, qualityReason: check.reason, qualityMeta: check.meta };
  }

  check = validateAiOutput(retry.text, quality);
  if (check.ok) {
    return { ...retry, retried: true, quality: check.meta };
  }

  return {
    ok: false,
    text: '',
    error: check.reason || 'quality_rejected',
    qualityMeta: check.meta,
    firstAttempt: first.text.slice(0, 400),
    secondAttempt: retry.text.slice(0, 400),
  };
}

module.exports = {
  generateText,
  generateChat,
  generateValidated,
  resolveOllamaBaseUrl,
  resolveModel,
  resolveTimeoutMs,
  resolveNumPredict,
  resolveTemperature,
  isEnabled,
};
